csi/src/error.rs

```diff
@@ -0,0 +1,290 @@
+    ArgumentInvalid {
+        /// Context of the error
+        context: Vec<String>,
+    },
+
+    /// Argument is out of range
+    #[error("Argument is out of range, context is {:?}", .context)]
+    ArgumentOutOfRange {
+        /// Context of the error
+        context: Vec<String>,
+    },
+
+    /// Error caused by std::path::StripPrefixError
+    #[error("StripPrefixErr, the error is {:?}, context is {:?}", .source, .context)]
+    StripPrefixErr {
+        /// Error source
+        source: std::path::StripPrefixError,
+        /// Context of the error
+        context: Vec<String>,
+    },
+
+    /// Error caused by etcd_client::Error,
+    #[error("EtcdClientErr, the error is {:?}, context is {:?}", .source, .context)]
+    EtcdClientErr {
+        /// Error source
+        source: etcd_client::Error,
+        /// Context of the error
+        context: Vec<String>,
+    },
+
+    /// Error caused by bincode::Error
+    #[error("BincodeErr, the error is {:?}, context is {:?}", .source, .context)]
+    BincodeErr {
+        /// Error source
+        source: bincode::Error,
+        /// Context of the error
+        context: Vec<String>,
+    },
+
+    /// Error caused by nix::Error
+    #[error("NixErr, the error is {:?}, context is {:?}", .source, .context)]
+    NixErr {
+        /// Error source
+        source: nix::Error,
+        /// Context of the error
+        context: Vec<String>,
+    },
+
+    /// Failed to mount
+    #[error("MountErr, fail to mount {:?} to {:?}, context is {:?}", .from, .target, .context)]
+    MountErr {
+        /// Source to mount
+        from: PathBuf,
+        /// Mount point
+        target: PathBuf,
+        /// Context of the error
+        context: Vec<String>,
+    },
+
+    /// Failed to umount
+    #[error("UmountErr, fail to umount {:?}, context is {:?}", .target, .context)]
+    UmountErr {
+        /// Mount point to umount
+        target: PathBuf,
+        /// Context of the error
+        context: Vec<String>,
+    },
+
+    /// Error caused by std::time::SystemTimeError
+    #[error("SystemTimeErr, the error is {:?}, context is {:?}", .source, .context)]
+    SystemTimeErr {
+        /// Error source
+        source: std::time::SystemTimeError,
+        /// Context of the error
+        context: Vec<String>,
+    },
+
+    /// Error caused by grpcio::Error
+    #[error("GrpcioErr, the error is {:?}, context is {:?}", .source, .context)]
+    GrpcioErr {
+        /// Error source
+        source: grpcio::Error,
+        /// Context of the error
+        context: Vec<String>,
+    },
+
+    /// API is not implemented
+    #[error("Not implemented, context is {:?}", .context)]
+    Unimplemented {
+        /// Context of the error
+        context: Vec<String>,
+    },
+}
+
+/// Add context to `DatenLordResult`
+pub trait Context<C, T> {
+    /// Add context to `DatenLordResult`
+    fn add_context(self, ctx: C) -> DatenLordResult<T>
+    where
+        C: Into<String>;
+}
```

Comment:
It's better to provide a lazy evaluated method. See [`anyhow::Context::with_context`](https://docs.rs/anyhow/1.0.38/anyhow/trait.Context.html#tymethod.with_context) and [`clippy::or_fun_call`](https://rust-lang.github.io/rust-clippy/master/index.html#or_fun_call).

Without lazy evaluation, we will pay for allocating and formatting lots of error strings even when everything is ok. It can be avoided by a simple method.

Reply:
Will add.